### PR TITLE
feat: [IDP-1825]: Regex support for filename

### PR DESCRIPTION
**Pull request by vigneswara-propelo**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56978)

### CODE REVIEW

**Review by vikyathharekal — COMMENTED** [review]
> Reviewed 19 of 27 files at r1, 1 of 1 files at r2, 8 of 8 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56978)* status: 29 file(s) to review. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/dsldataprovider/impl/BitbucketContentDsl.java` line 83 at r4](https://reviewable.io/reviews/harness/harness-core/56978#-Nm4upll9fOLTbrg5fX_:-Nm4upll9fOLTbrg5fXa:bx0ytom) ([raw file](https://github.com/harness/harness-core/blob/18c9ab1fe040ec7700ca10204fb731441cb1a3ca/idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/dsldataprovider/impl/BitbucketContentDsl.java#L83)):*
> ```Java
>       Pattern pattern = getFileNamePattern(scmConfig);
>       String matchingFile = null;
>       for (Map<String, Object> value : values) {
> ```

Does the API give the list of files in a certain order? We might have to make it clear in docs as to what happens when there is more than one file matching the pattern



_Code quote:_
```Java
values
```
___
*[`idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/dsldataprovider/impl/GithubContentDsl.java` line 74 at r4](https://reviewable.io/reviews/harness/harness-core/56978#-Nm4vn4D3cfW022yAz_7:-Nm4vn4D3cfW022yAz_8:bbwmh9r) ([raw file](https://github.com/harness/harness-core/blob/18c9ab1fe040ec7700ca10204fb731441cb1a3ca/idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/dsldataprovider/impl/GithubContentDsl.java#L74)):*
> ```Java
>     if (response.getStatus() == 200) {
>       Map<String, Object> data = GsonUtils.convertJsonStringToObject(response.getEntity().toString(), Map.class);
>       if (CommonUtils.findObjectByName(data, "object") != null) {
> ```

In case of non-200 status or if we don't find "object", we return null. Does this get propagated to the UI?



_Code quote:_
```Java
    if (response.getStatus() == 200) {
      Map<String, Object> data = GsonUtils.convertJsonStringToObject(response.getEntity().toString(), Map.class);
      if (CommonUtils.findObjectByName(data, "object") != null) {
```
___
*[`idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/dsldataprovider/impl/GitlabContentDsl.java` line 48 at r4](https://reviewable.io/reviews/harness/harness-core/56978#-Nm4wrz0BFCI29oyvuY8:-Nm4wrz0BFCI29oyvuY9:b-d528c7) ([raw file](https://github.com/harness/harness-core/blob/18c9ab1fe040ec7700ca10204fb731441cb1a3ca/idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/dsldataprovider/impl/GitlabContentDsl.java#L48)):*
> ```Java
> 
>   @Override
>   public Map<String, Object> getDslData(String accountIdentifier, Object config) {
> ```

Can we check if this method can be generalized a bit? We have similar code in all 3 DSLs.
Same with getFileName() method



_Code quote:_
```Java
getDslData
```
___
*[`idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/resource/ScmDataPointsApiImpl.java` line 32 at r4](https://reviewable.io/reviews/harness/harness-core/56978#-Nm4xdeU28-H3LaZUdLk:-Nm4xdeU28-H3LaZUdLl:b-audmzl) ([raw file](https://github.com/harness/harness-core/blob/18c9ab1fe040ec7700ca10204fb731441cb1a3ca/idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/resource/ScmDataPointsApiImpl.java#L32)):*
> ```Java
>   ScmDataPointsService scmDataPointsService;
>   @Override
>   public Response getScmDataPointValues(@Valid ScmRequest body, String harnessAccount) {
> ```

nit: @AccountIdentifier annotation



_Code quote:_
```Java
harnessAccount
```
___
*[`idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/service/ScmDataPointsServiceImpl.java` line 36 at r4](https://reviewable.io/reviews/harness/harness-core/56978#-Nm4xyGk7-KzahKuuPsP:-Nm4xyGk7-KzahKuuPsQ:bj84quo) ([raw file](https://github.com/harness/harness-core/blob/18c9ab1fe040ec7700ca10204fb731441cb1a3ca/idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/service/ScmDataPointsServiceImpl.java#L36)):*
> ```Java
> 
>   @Override
>   public Map<String, Object> getDataPointDataValues(String accountIdentifier, ScmConfig scmConfig) {
> ```

Please check if this can be made generic for K8s, Harness and Scm



_Code quote:_
```Java
getDataPointDataValues
```
___
*[`idp-service/src/main/java/io/harness/idp/scorecard/datasourcelocations/locations/scm/ScmProxyThroughDsl.java` line 45 at r4](https://reviewable.io/reviews/harness/harness-core/56978#-Nm4zNle16C35VN-Xo0A:-Nm4zNle16C35VN-Xo0B:b-7nfilj) ([raw file](https://github.com/harness/harness-core/blob/18c9ab1fe040ec7700ca10204fb731441cb1a3ca/idp-service/src/main/java/io/harness/idp/scorecard/datasourcelocations/locations/scm/ScmProxyThroughDsl.java#L45)):*
> ```Java
>   @Inject @Named("env") private String env;
>   @Inject @Named("base") private String base;
>   ScmConfig scmConfig = new ScmConfig();
> ```

Having this as class variable could have side effects. Like REPOSITORY_SUB_FOLDER is conditional, so we might end up using previous call's value in the current one.



_Code quote:_
```Java
ScmConfig scmConfig = new ScmConfig();
```
___
*[`idp-service/src/main/resources/migrations/scorecard/datasourceLocations.json` line 425 at r4](https://reviewable.io/reviews/harness/harness-core/56978#-Nm5-fVX7doJNL5meT2g:-Nm5-fVX7doJNL5meT2h:bftqh65) ([raw file](https://github.com/harness/harness-core/blob/18c9ab1fe040ec7700ca10204fb731441cb1a3ca/idp-service/src/main/resources/migrations/scorecard/datasourceLocations.json#L425)):*
> ```JSON
>         "Authorization": "{SERVICE_TO_SERVICE_AUTH}",
>         "Harness-Account": "{ACCOUNT_IDENTIFIER}",
>         "X-Source-Principal": "{X_SOURCE_PRINCIPAL}"
> ```

we can move these common headers to dataSources.json



_Code quote:_
```JSON
        "Authorization": "{SERVICE_TO_SERVICE_AUTH}",
        "Harness-Account": "{ACCOUNT_IDENTIFIER}",
        "X-Source-Principal": "{X_SOURCE_PRINCIPAL}"
```

**Review by vigneswara-propelo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56978)* status: 38 file(s) to review. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/dsldataprovider/impl/BitbucketContentDsl.java` line 83 at r4](https://reviewable.io/reviews/harness/harness-core/56978#-Nm4upll9fOLTbrg5fX_:-Nm68Y4LDPfc_wggPWmp:b-ob5uiq) ([raw file](https://github.com/harness/harness-core/blob/18c9ab1fe040ec7700ca10204fb731441cb1a3ca/idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/dsldataprovider/impl/BitbucketContentDsl.java#L83)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Does the API give the list of files in a certain order? We might have to make it clear in docs as to what happens when there is more than one file matching the pattern
</blockquote></details>

Yeah, all three providers list filenames in ascending order and we will be looking for the first matching file. We can mention the same in the docs 
___
*[`idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/dsldataprovider/impl/GithubContentDsl.java` line 74 at r4](https://reviewable.io/reviews/harness/harness-core/56978#-Nm4vn4D3cfW022yAz_7:-Nm68le--VG7y6CysJwb:btc0d21) ([raw file](https://github.com/harness/harness-core/blob/18c9ab1fe040ec7700ca10204fb731441cb1a3ca/idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/dsldataprovider/impl/GithubContentDsl.java#L74)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

In case of non-200 status or if we don't find "object", we return null. Does this get propagated to the UI?
</blockquote></details>

1. status != 200 -> processResponse method will handle different scenarios 
2. Modified code to validate "object" != null case
___
*[`idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/dsldataprovider/impl/GitlabContentDsl.java` line 48 at r4](https://reviewable.io/reviews/harness/harness-core/56978#-Nm4wrz0BFCI29oyvuY8:-Nm698X3AfNhk3XV2nB7:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/18c9ab1fe040ec7700ca10204fb731441cb1a3ca/idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/dsldataprovider/impl/GitlabContentDsl.java#L48)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Can we check if this method can be generalized a bit? We have similar code in all 3 DSLs.
Same with getFileName() method
</blockquote></details>

Done.
___
*[`idp-service/src/main/java/io/harness/idp/scorecard/datasourcelocations/locations/scm/ScmProxyThroughDsl.java` line 45 at r4](https://reviewable.io/reviews/harness/harness-core/56978#-Nm4zNle16C35VN-Xo0A:-Nm69CWv8puFlZkdohha:bevgaqj) ([raw file](https://github.com/harness/harness-core/blob/18c9ab1fe040ec7700ca10204fb731441cb1a3ca/idp-service/src/main/java/io/harness/idp/scorecard/datasourcelocations/locations/scm/ScmProxyThroughDsl.java#L45)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Having this as class variable could have side effects. Like REPOSITORY_SUB_FOLDER is conditional, so we might end up using previous call's value in the current one.
</blockquote></details>

Thanks, missed to check this. Now initialisation will be done in the matchAndReplaceHeaders() 
___
*[`idp-service/src/main/resources/migrations/scorecard/datasourceLocations.json` line 425 at r4](https://reviewable.io/reviews/harness/harness-core/56978#-Nm5-fVX7doJNL5meT2g:-Nm69lJe8kBJgURCRBEd:b2cxh71) ([raw file](https://github.com/harness/harness-core/blob/18c9ab1fe040ec7700ca10204fb731441cb1a3ca/idp-service/src/main/resources/migrations/scorecard/datasourceLocations.json#L425)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

we can move these common headers to dataSources.json
</blockquote></details>

For scm providers, we already have "Authorization": "{BEARER_AUTH}" and it would conflict each other if we keep it there

**Review by vigneswara-propelo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56978)* status: 38 file(s) to review. 6. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/service/ScmDataPointsServiceImpl.java` line 36 at r4](https://reviewable.io/reviews/harness/harness-core/56978#-Nm4xyGk7-KzahKuuPsP:-Nm6AGxp7fB8GVtMQGQj:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/18c9ab1fe040ec7700ca10204fb731441cb1a3ca/idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/service/ScmDataPointsServiceImpl.java#L36)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Please check if this can be made generic for K8s, Harness and Scm
</blockquote></details>

Done.

**Review by mankrit-singh — APPROVED** [review]

File: idp-service/src/main/resources/migrations/scorecard/datasourceLocations.json [hunk 1]
```diff
@@ -447,11 +451,14 @@
     "dataSourceIdentifier": "bitbucket",
     "type": "DirectHttp",
```
Comment (sathish-soundarapandian) [inline idp-service/src/main/resources/migrations/scorecard/datasourceLocations.json:452]: This should be custom http ?

**Review by pranay-harness — APPROVED** [review]

File: idp-service/src/main/resources/migrations/scorecard/datasourceLocations.json (hunk 1 above)
Comment (vigneswara-propelo) [reply idp-service/src/main/resources/migrations/scorecard/datasourceLocations.json:452]: yeah, missed to add. Will raise a separate PR for this one line change

**Review by vikyathharekal — APPROVED** [review]
> Reviewed 17 of 18 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56978)* status: 38 file(s) to review. 2. unresolved discussions. 

___
*[`idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/dsldataprovider/impl/ScmContentDsl.java` line 44 at r6](https://reviewable.io/reviews/harness/harness-core/56978#-Nm6orny5OLbZyN5_IEj:-Nm6orny5OLbZyN5_IEk:b-u4vpch) ([raw file](https://github.com/harness/harness-core/blob/3e2e4314798b8b89dcaaaccff01c9e3536335c63/idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/dsldataprovider/impl/ScmContentDsl.java#L44)):*
> ```Java
> @OwnedBy(HarnessTeam.IDP)
> public abstract class ScmContentDsl {
>   String FILE_PATH_REPLACER = "{FILE_PATH_REPLACER}";
> ```

nit: we can make it static final



_Code quote:_
```Java
String
```
___
*[`idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/service/DataPointDataValueServiceImpl.java` line 61 at r6](https://reviewable.io/reviews/harness/harness-core/56978#-Nm6shQp2G6hD2YDADMT:-Nm6shQp2G6hD2YDADMU:b-eb23ta) ([raw file](https://github.com/harness/harness-core/blob/3e2e4314798b8b89dcaaaccff01c9e3536335c63/idp-service/src/main/java/io/harness/idp/scorecard/datapointsdata/service/DataPointDataValueServiceImpl.java#L61)):*
> ```Java
>       case HARNESS_IDENTIFIER:
>         return ((DataSourceDataPointInfo) config)
>             .getDataSourceLocation()
> ```

nit: we could get the dsl object from each case and do the remaining at the end in common code.



_Code quote:_
```Java
.getDataSourceLocation()
```

**Comment by vigneswara-propelo** [comment]
trigger review

**Comment by vigneswara-propelo** [comment]
trigger ut5, ut6, ut7, ut8, ut9

**Comment by vigneswara-propelo** [comment]
trigger ut5, ut6, ut7, ut8, ut9

**Comment by vigneswara-propelo** [comment]
trigger ut5, ut6

**Pull request merged**